test/kafka-upsert-benchmark/mzworkflows.py

```diff
@@ -0,0 +1,165 @@
+    name="materialized_persist",
+    hostname="materialized",
+    options="--persistent-user-tables --persistent-kafka-upsert-source --disable-persistent-system-tables-test",
+)
+
+mz_default = Materialized()
+
+prerequisites = ["zookeeper", "kafka"]
+
+services = [
+    Zookeeper(),
+    Kafka(),
+    mz_default,
+    mz_persist,
+    # This instance of Testdrive is used to initialize the benchmark, and needs to have a
+    # known seed.
+    Testdrive(no_reset=True, seed=1),
+    Kgen(),
+]
+
+
+def kafka_upsert_benchmark_step(
+    w: Workflow,
+    mz_service: str,
+    restart_mz: bool,
+    records_sent: int,
+    num_records: int,
+    num_keys: int,
+    value_bytes: int,
+    timeout_secs: int,
+):
+    if restart_mz:
+        w.kill_services(services=[mz_service], signal="SIGKILL")
+    w.run_service(
+        service="kgen",
+        command=" ".join(
+            [
+                f"--num-records={num_records}",
+                "--keys=random",
+                "--key-min=0",
+                f"--key-max={num_keys}",
+                "--values=bytes",
+                f"--max-message-size={value_bytes+1}",
+                f"--min-message-size={value_bytes}",
+                "--topic=testdrive-load-test-1",
+                "--quiet",
+            ]
+        ),
+    )
+
+    if restart_mz:
+        w.start_services(services=[mz_service])
+        w.wait_for_mz(service=mz_service)
+    expected = records_sent + num_records
+
+    host_port = int(w.composition.find_host_ports(mz_service)[0])
+    query = "SELECT * FROM num_records_ingested"
+    ui.progress(f"waiting for materialize to handle {query!r}", "C")
+    error = None
+    start_time = time.monotonic()
+
+    # Default to polling Materialize once a second because that is the current default
+    # timestamp frequency and polling any more frequently doesn't give new data.
+    for remaining in ui.timeout_loop(timeout_secs):
+        try:
+            conn = pg8000.connect(
+                host="localhost", port=host_port, user="materialize", timeout=1
+            )
+            conn.autocommit = True
+
+            cursor = conn.cursor()
+            cursor.execute("SELECT * FROM num_records_ingested")
+            n = cursor.fetchone()[0]
+            curr_time = time.monotonic()
+            elapsed = curr_time - start_time
+            if n == expected:
+                say(f"query result: {n} after {elapsed}")
+                ui.progress("success!", finish=True)
+                return
+            else:
+                say(f"materialize has not ingested {expected} records, got: {n}")
+        except Exception as e:
+            ui.progress(" " + str(int(remaining)))
+            error = e
+    ui.progress(finish=True)
+    raise UIError(f"never got correct result {expected}: {error}")
+
+
+def workflow_kafka_upsert_benchmark(w: Workflow, args: List[str]):
+    parser = WorkflowArgumentParser(w)
+    parser.add_argument("--num-steps", type=int, default=10)
```

Comment:
With those defaults, the benchmark is measuring the time the `SELECT` took to run, as the data is fully ingested by the time the statement executes. So the numbers are into the low milliseconds range:

```
C> waiting for materialize to handle 'SELECT * FROM num_records_ingested'C> query result: 200000 after 0.004557367414236069
```

You may wish to consider much higher defaults, ones that will push the results far beyond the 1-second granularity of the measurements.

Reply:
changed the approach to be totally different here, but I agree with that analysis. had picked the defaults to be low to not surprise first time users - not a very good reason.